docs/spec/escrow.rst

```diff
@@ -0,0 +1,312 @@
+==============  ============  ========================================================
+
+Functions
+~~~~~~~~~
+
+.. _escrow-function-create-lock:
+
+create_lock
+-----------
+
+Create a lock on the account's balance to expire in the future.
+
+Specification
+.............
+
+*Function Signature*
+
+``create_lock(who, amount, unlock_height)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``amount``: The amount to be locked.
+* ``unlock_height``: The height to lock until.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account ``who`` MUST NOT already have locked balance.
+* The ``unlock_height`` MUST be greater than ``now``.
+* The ``unlock_height`` MUST NOT be greater than ``now + MaxPeriod``.
+
+*Postconditions*
+
+* The account's ``LockedBalance`` MUST be set as follows:
+
+    * ``new_locked.amount``: MUST be the ``amount``.
+    * ``new_locked.end``: MUST be the ``unlock_height``.
+
+.. _escrow-function-increase-amount:
+
+increase_amount
+---------------
+
+Deposit additional tokens for a pre-existing lock to improve voting power.
+
+Specification
+.............
+
+*Function Signature*
+
+``increase_amount(who, amount)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``amount``: The amount to be locked.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account's ``old_locked.amount`` MUST be non-zero.
+* The account's ``old_locked.end`` MUST be greater than ``now``.
+
+*Postconditions*
+
+* The account's ``LockedBalance`` MUST be set as follows:
+
+    * ``new_locked.amount``: MUST be ``old_locked.amount + amount``.
+    * ``new_locked.end``: MUST be the ``old_locked.end``.
+
+.. _escrow-function-increase-unlock-height:
+
+increase_unlock_height
+----------------------
+
+Push back the expiry on a pre-existing lock to retain voting power.
+
+Specification
+.............
+
+*Function Signature*
+
+``increase_unlock_height(who, unlock_height)``
+
+*Parameters*
+
+* ``who``: The user's address.
+* ``unlock_height``: The new expiry deadline.
+
+*Events*
+
+* :ref:`escrow-event-deposit`
+
+*Preconditions*
+
+* The function call MUST be signed by ``who``.
+* The ``amount`` MUST be non-zero.
+* The account's ``old_locked.amount`` MUST be non-zero.
+* The account's ``old_locked.end`` MUST be greater than ``now``.
+* The ``unlock_height`` MUST be greater than ``old_locked.end``.
+* The ``unlock_height`` MUST NOT be greater than ``now + MaxPeriod``.
+
+*Postconditions*
+
+* The account's ``LockedBalance`` MUST be set as follows:
+
+    * ``new_locked.amount``: MUST be ``old_locked.amount``.
+    * ``new_locked.end``: MUST be the ``unlock_height``.
+
+.. _escrow-function-withdraw:
+
+increase_withdraw
+-----------------
+
+Remove the lock on an account.
```

Comment:
Should make clear that only tokens that are unlocked can be withdrawn. 

Also: do we allow withdraw(amount) or is there a reason for not doing this?

Reply:
The way that the function is currently defined, as I understand it, you can only withdraw once the entire lock has passed. So e.g. you lock 100 tokens at t_0 until t_100. At t_50 you cannot withdraw any tokens since you committed to locking them until t_100. So you unlock 100% (after t_100) or none at all (before t_100). Which I think is in line with how the system is supposed to work. For the same reason, `withdraw(amount)` should not be possible.

I guess @gregdhill had the same understanding. That's also how Curve works: 
- https://curve.readthedocs.io/dao-vecrv.html#working-with-vote-locks
- https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/VotingEscrow.vy#L469
It's only possible to withdraw once the lock expires for all tokens.

Reply:
@nud3l is correct, we can only withdraw the full amount after the lock has expired.